drivers/i2s/i2s_nrfx.c

```diff
@@ -0,0 +1,1193 @@
+			return -EIO;
+		}
+		channel_change_state(channel, I2S_STATE_NOT_READY);
+		channel_mem_clear(i2s, dir);
+		return 0;
+	}
+
+	ret = cfg_periph_config(dev, &drv_cfg, i2s_cfg);
+	/* disable channels in case of invalid configuration*/
+	if (ret < 0) {
+		LOG_ERROR("Config: Failed to configure peripheral");
+		channel_change_state(channel, I2S_STATE_ERROR);
+		return ret;
+	}
+	if (dir == I2S_DIR_RX) {
+		i2s->buffers.p_rx_buffer = NULL;
+	}
+	if (dir == I2S_DIR_TX) {
+		i2s->buffers.p_tx_buffer = NULL;
+	}
+	channel->mem_slab = i2s_cfg->mem_slab;
+	channel->timeout = i2s_cfg->timeout;
+	ret = channel_change_state(channel, I2S_STATE_READY);
+	if (ret < 0) {
+		channel_change_state(channel, I2S_STATE_ERROR);
+		return ret;
+	}
+
+	if (interface_get_state(i2s) == I2S_IF_NOT_READY) {
+		status = nrfx_i2s_init(&drv_cfg, interface_handler);
+		if (status != NRFX_SUCCESS) {
+			if (status == NRFX_ERROR_INVALID_STATE) {
+				LOG_ERROR("Config: NRFX Invalid state");
+				channel_change_state(channel, I2S_STATE_ERROR);
+				return -EINVAL;
+			} else if (status == NRFX_ERROR_INVALID_PARAM) {
+				LOG_ERROR("Config: NRFX Invalid parameter");
+				channel_change_state(channel, I2S_STATE_ERROR);
+				return -EINVAL;
+			}
+			LOG_ERROR("Config: NRFX General error");
+			channel_change_state(channel, I2S_STATE_ERROR);
+			return -ENOTSUP;
+		}
+		ret = interface_set_state(i2s, I2S_IF_READY);
+		if (ret < 0) {
+			channel_change_state(channel, I2S_STATE_ERROR);
+			return -ENOTSUP;
+		}
+	} else if (interface_get_state(i2s) != I2S_IF_READY) {
+		LOG_ERROR("Config: Interface must be in ready state");
+		channel_change_state(channel, I2S_STATE_ERROR);
+		return -ENOTSUP;
+	}
+	memcpy(&channel->config, i2s_cfg, sizeof(struct i2s_config));
+	return ret;
+}
+
+static struct i2s_config *i2s_nrfx_config_get(struct device *dev,
+					     enum i2s_dir dir)
+{
+	struct i2s_nrfx_data *i2s = get_interface();
+
+	assert(dev != NULL);
+	return &channel_get(i2s, dir)->config;
+}
+
+static int i2s_nrfx_read(struct device *dev, void **mem_block, size_t *size)
+{
+	struct i2s_nrfx_data *i2s = get_interface();
+	struct channel_str *ch_rx = &i2s->channel_rx;
+	int ret;
+
+	assert(dev != NULL && mem_block != NULL && size != NULL);
+	*size = 0;
+	if (ch_rx->current_state == I2S_STATE_NOT_READY ||
+	    ch_rx->current_state == I2S_STATE_ERROR) {
+		return -EIO;
+	}
+	ret = k_sem_take(&ch_rx->sem, ch_rx->timeout);
+	if (ret < 0) {
+		return ret;
+	}
+	ret = queue_fetch(&ch_rx->mem_block_queue, mem_block);
+	if (ret < 0) {
+		return -EIO;
+	}
+	*size = i2s->size;
+	return 0;
+}
+
+static int i2s_nrfx_write(struct device *dev, void *mem_block, size_t size)
+{
+	struct i2s_nrfx_data *i2s = get_interface();
+	struct channel_str *ch_tx = &i2s->channel_tx;
+	int ret;
+
+	assert(dev != NULL && mem_block != NULL);
+	if (ch_tx->current_state != I2S_STATE_READY &&
+	    ch_tx->current_state != I2S_STATE_RUNNING) {
+		return -EIO;
+	}
+	if (size != i2s->size) {
```

Comment:
According to the I2S API: @param size Number of bytes to write. This value has to be equal or smaller
than the size of the memory block. Seems like this check will prevent size from being smaller than the mem block.

Reply:
Nordic I2S peripheral doesn't allow to change data size during transfer. Consequently this function should return error in case of try to send data size other than declared during configuration.